### PR TITLE
Record stripe charge data through a webhook

### PR DESCRIPTION
**Pull request by cadecairos**
Fixes mozilla/donate.mozilla.org#623

Still a WIP, but mostly implemented

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,6 +27,19 @@ CREATE TABLE paypal (
   currency TEXT,
   country_code TEXT
 );
+
+CREATE TABLE stripe (
+  id TEXT PRIMARY KEY,
+  timestamp TIMESTAMP WITH TIME ZONE,
+  amount INTEGER,
```
Comment (jbuck) [inline README.md]: Could you change this to `MONEY`?

Comment (cadecairos) [inline README.md]: The `MONEY` type assumes it's not a 0-decimal currency, so I don't think thats a good idea.

`settle_amount`, just below this line, is a `MONEY` type and reflects the actual amount dontated in USD at the time of the charge.

Keeping this an integer ensures we can convert the zero decimal based currencies in the future if we need to.

Comment (jbuck) [inline README.md]: I don't think it should be an integer though. What if I donate $10.10 USD? We'd lose the $0.10 precision

Comment (cadecairos) [inline README.md]: if the transaction is in USD, stripe reports the value in cents, so 10.10 would be 1010 cents

**Comment by cadecairos** [comment]
@jbuck updated - R?

File: lib/stripe.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    client.query(upsert_query, [
+      charge.id,
+      moment.unix(charge.created).toISOString(),
+      adjusted_amount,
+      settle_amount,
+      adjusted_amount_refunded,
+      currency,
+      charge.status,
+      charge.source.address_country,
+      charge.name,
```
Comment (jbuck) [inline lib/stripe.js]: This should be `charge.source.name`

File: stripe.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+var process_charge = require('./lib/stripe').process_charge;
+
+module.exports = function(request, reply) {
+ // stripe doesn't care about the result of this request, so end the request.
```
Comment (jbuck) [inline stripe.js]: I'd wait until process_charge has completed successfully before `reply()`-ing to Stripe. That way Stripe will repeat the webhook if we fail to respond with a 2xx response

File: stripe.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  reply();
+
+  var event = request.payload;
+
+if (event && event.data && event.data.object) {
+    // verify that this is a real stripe charge associated with our stripe account.
```
Comment (jbuck) [inline stripe.js]: The logic to do this should be added somewhere, or add some sort of access control to the server

**Comment by jbuck** [comment]
Some more work needs to be done, but a fine start!

**Comment by cadecairos** [comment]
@jbuck updated!

File: README.md [hunk 5]
```diff
@@ -43,3 +56,7 @@ Create a file named `.env` to place configuration options in.
 * SERVER_DB_CONNECTION_STRING
 * SERVER_START_DATE
 * SERVER_END_DATE
+* SERVER_STRIPE_SECRET
```
Comment (jbuck) [inline README.md]: Isn't this the same as the STRIPE_API_KEY?

Comment (cadecairos) [inline README.md]: I don't think we should use the stripe API key as our bearer token for the webhook endpoint

Comment (jbuck) [inline README.md]: hahahaha oh boy. Sorry. Yes

File: stripe.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+Configure stripe to only hit this endpoint with the following events:
+- charge.succeeded
+- charge.captured
+- charge.failed
+- charge.refunded
+- charge.updated
+*/
+
+module.exports = function(request, reply) {
+  var event = request.payload;
```
Comment (jbuck) [inline stripe.js:14]: Could you add a check here that the event type you're receiving is the type you expect? And just reply() immediately if it isn't.

Comment (cadecairos) [inline stripe.js:14]: done!

**Comment by jbuck** [comment]
belated r+. gourd work @cadecairos 

Lets make sure the data is all good in the morning

**Pull request merged**